bigquery/storage_session.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bigquery
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	gax "github.com/googleapis/gax-go/v2"
+	"google.golang.org/api/iterator"
+	bqStoragepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1"
+	storagepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1"
+	"google.golang.org/grpc"
+)
+
+// ReadSession is the abstraction over a storage API read session.
+type ReadSession struct {
+	settings *settings
+	rc       *ReadClient
+
+	ctx     context.Context
+	table   *Table
+	tableID string
+
+	bqSession *bqStoragepb.ReadSession
+
+	// SerializedArrowSchema is an IPC-serialized Arrow Schema.
```

Comment:
Rather than exposing all these mutable things on the ReadSession, should it instead just have a method which returns a proto.Clone if the retained ReadSession?

Reply:
I like the idea of returning a clone to avoid users messing with the attributes as they are used locally, but should we return our own `ReadSession` or the `bqStoragepb.ReadSession`  ? The second option would give access to internal proto stubs from the Storage API.
```
bqStoragepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1"
```

Reply:
Yeah, we could certainly return a custom type with the exported fields as well.